tensorflow-examples/src/main/java/org/tensorflow/model/examples/cnn/fastrcnn/FasterRcnnInception.java

```diff
@@ -101,250 +101,144 @@ The given SavedModel SignatureDef contains the following output(s):
+            .output("image", image)
+            .build();
+    }
+
+    private static Signature writeOutputImage(Ops tf) {
+        Placeholder<TString> outputImagePath = tf.placeholder(TString.class);
+        Placeholder<TUint8> inputImage = tf.placeholder(TUint8.class);
+        Placeholder<TFloat32> detectionBoxes = tf.placeholder(TFloat32.class);
+
+        //2-D. A list of RGBA colors to cycle through for the boxes.
+        Operand<TFloat32> colors = tf.constant(new float[][]{
+                {0.9f, 0.3f, 0.3f, 0.0f},
+                {0.3f, 0.3f, 0.9f, 0.0f},
+                {0.3f, 0.9f, 0.3f, 0.0f}
+        });
+        //convert the 4D input image to normalised 0.0f - 1.0f
+        //Draw bounding boxes using boxes tensor and list of colors
+        //multiply by 255 then reshape and recast to TUint8 3D tensor
+        tf.io.writeFile(
+            outputImagePath,
+            tf.image.encodeJpeg(
+                    tf.dtypes.cast(
+                            tf.squeeze(
+                                    tf.math.mul(
+                                            tf.image.drawBoundingBoxes(
+                                                    tf.math.div(
+                                                            tf.dtypes.cast(inputImage, TFloat32.class),
                                                             tf.constant(255.0f)
                                                     ),
-                                                    tf.array(
-                                                            imageShape.asArray()[0],
-                                                            imageShape.asArray()[1],
-                                                            imageShape.asArray()[2]
-                                                    )
-                                            ), TUint8.class),
-                                            jpgOptions));
-                            //output the JPEG to file
-                            s.runner().feed(outImagePathPlaceholder, TString.scalarOf(outputImagePath))
-                                    .feed(boxesPlaceHolder, boxes)
-                                    .addTarget(writeFile).run();
-                        }
-                    }
-                }
-            }
+                                                    detectionBoxes,
+                                                    colors
+                                            ),
+                                            tf.constant(255.0f)
+                                    )
+                            ),
+                            TUint8.class
+                    ),
+                    EncodeJpeg.quality(100L) //Create JPEG from the Tensor with quality of 100%
+            )
+        );
+
+        return Signature.builder()
+            .input("outputImagePath", outputImagePath)
+            .input("inputImage", inputImage)
+            .input("boxes", detectionBoxes)
```

Comment:
Doesn't this need to be connected to the writeFile op in some way?

Reply:
The pre/post processing is executed eagerly, if that's what you are wandering?

Reply:
Aren't the two concrete functions created as graphs? If so I don't understand how the writefile op is executed as while the graph contains it it's not attached to any target & the buildFromGraph logic traces all the ops between the input and the output.

Reply:
No, the `tf` the receive in parameters is coming from the [default eager session](https://github.com/tensorflow/java/blob/50f7efa5821877668c2645886b0e4addef7306b6/tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/ConcreteFunction.java#L303). That's the behavior when you are calling a function directly (i.e. `function.call(params)`)

Reply:
`ConcreteFunction.create(Function<Ops,Signature>)` traces the function by adding it to a graph and asking for all the ops between the inputs & outputs, but as there aren't any outputs in the signature I don't think it'll add any ops bar the inputs to the function when it's created.

Reply:
Oh you are right, so I think the dark magic here is that the function will result in a graph but will be executed eagerly, i.e. right away, the default eager session is used for that.

I know it works the way it is because I got the output image written on my side. Now how does it link to `writeFile`, mmmh... I'm now confused as well. But it works ;)